### PR TITLE
Testing Overview: update JS example to use React Testing Library

### PR DESCRIPTION
**Pull request by ciampo**
## Description

As agreed in https://github.com/WordPress/gutenberg/pull/33960#discussion_r695118872, this PR updates the JS section of the `Testing Overview` docs:

- Enzyme examples refactored to React Testing Library
- Slightly changed the wording around RTL and Enzyme
- Added a section about the `toMatchDiffSnapshot` function

## How has this been tested?

👀 

## Types of changes

Docs

## Checklist:
- N/A My code is tested.
- N/A My code follows the WordPress code style. 
- N/A My code follows the accessibility standards. 
- N/A I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- N/A I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: docs/contributors/code/testing-overview.md
Comment (diegohaz) [inline docs/contributors/code/testing-overview.md]: Let's recommend `screen` instead:
```jsx
import { render, screen } from '@testing-library/react';

const { container } = render( <SolarSystem planets /> );

expect( screen.getByText( /mars/i ) ).toBeInTheDocument();
```

More info: https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen

File: docs/contributors/code/testing-overview.md
Comment (diegohaz) [inline docs/contributors/code/testing-overview.md]: Are the extra lines intentional?
```suggestion
	} );
} );
```

File: docs/contributors/code/testing-overview.md
Comment (diegohaz) [inline docs/contributors/code/testing-overview.md]:
```suggestion
Another good technique is to use the `toMatchDiffSnapshot` function (provided by the [`snapshot-diff` package](https://github.com/jest-community/snapshot-diff)), which allows to snapshot only the difference between two different states of the DOM. This approach is useful to test the effects of a prop change on the resulting DOM while generating a much smaller snapshot, like in this example:
```

File: docs/contributors/code/testing-overview.md
Comment (diegohaz) [inline docs/contributors/code/testing-overview.md]:
```suggestion
```jsx
test( 'should render a darker background when isShady is true', () => {
```

File: docs/contributors/code/testing-overview.md
Comment (diegohaz) [inline docs/contributors/code/testing-overview.md]:
```suggestion
```jsx
```

File: docs/contributors/code/testing-overview.md
Comment (diegohaz) [inline docs/contributors/code/testing-overview.md]:
```suggestion
```jsx
```

File: docs/contributors/code/testing-overview.md
Comment (diegohaz) [reply docs/contributors/code/testing-overview.md]: I don't know if the handbook supports ```jsx. In case it doesn't, just ignore these suggestions.

**Comment by sarayourfriend** [comment]
In addition to `toMatchDiffSnapshot` should there be a blurb about `toMatchStyleDiffSnapshot`?

File: docs/contributors/code/testing-overview.md
Comment (gziolo) [reply docs/contributors/code/testing-overview.md]: Good idea, we use it in a few places already. I read the same blog post while doing some refactor a while ago 👍🏻

File: docs/contributors/code/testing-overview.md
Comment (gziolo) [reply docs/contributors/code/testing-overview.md]: Let's give it a try and see if it works. We can revert to `js` in a follow-up PR if the handbook finds it problematic.

File: docs/contributors/code/testing-overview.md
Comment (gziolo) [inline docs/contributors/code/testing-overview.md]: The main point in this note is to ensure that whenever Enzyme is used, it should be refactored to RTL if possible. 
```suggestion
_It should be noted that in the past, React components were unit tested with [Enzyme](https://github.com/airbnb/enzyme). However, React Testing Library (RTL) should be used for new tests instead, and over time old tests should be refactored to use RTL too (typically when working on code that touches an old test)._
```

File: docs/contributors/code/testing-overview.md
Comment (gziolo) [inline docs/contributors/code/testing-overview.md]: Is this change necessary? Is there anything wrong with using React testing tools directly?

**Review by gziolo — APPROVED** [review]
There are some tiny details to discuss before this lands. 

Anyway, overall these changes look great. Thank you very much for bringing so desperately needed refresh to testing overview ❤️

File: docs/contributors/code/testing-overview.md
Comment (diegohaz) [reply docs/contributors/code/testing-overview.md]: @gziolo Do you mean `react-test-renderer` and `react-dom/test-utils`? I would highly recommend against `react-test-renderer`, which is worse than Enzyme. `react-dom/test-utils` could work for simple scenarios, but it's ultra verbose compared to RTL. You can see a comparison between different testing approaches here: https://github.com/WordPress/gutenberg/issues/17249#issuecomment-544337655

Comment (gziolo) [reply docs/contributors/code/testing-overview.md]: React still keeps both those libraries/utils documented: 
https://reactjs.org/docs/test-utils.html

At the same time, they recommend RTL which isn't as strong as the phrase "should use". In addition to that, `act` from RTL is a simple wrapper for `react-dom/test-utils`. So there is still value in those primitives that React team provided. My main concern is that enforcing one library will impact the review process where the quality of tests would become less important than its internal implementation.

File: docs/contributors/code/testing-overview.md
Comment (ciampo) [reply docs/contributors/code/testing-overview.md]: Yeah, absolutely — I actually remember reading that post too. Thank you for the suggestion, diego!

File: docs/contributors/code/testing-overview.md
Comment (ciampo) [reply docs/contributors/code/testing-overview.md]: Those are all fair points!

My bad, I initially thought that we'd want to push for RTL with a bit more than a recommendation. I will undo the changes and leave that paragraph untouched.

Comment (gziolo) [reply docs/contributors/code/testing-overview.md]: We can add more recommendations to help people make the right choice which we all agree that RTL is. We can also add an ESLint rule that errors when someone adds `enzyme` in the test file (we would have to add exceptions in existing test files though).

File: docs/contributors/code/testing-overview.md
Comment (ciampo) [reply docs/contributors/code/testing-overview.md]: Done in [`96aa0c7` (#34423)](https://github.com/WordPress/gutenberg/pull/34423/commits/96aa0c72458d286589c07015ac3742c35a968858)

File: docs/contributors/code/testing-overview.md
Comment (ciampo) [reply docs/contributors/code/testing-overview.md]: >  I will undo the changes and leave that paragraph untouched.

Done in [`5dc8719` (#34423)](https://github.com/WordPress/gutenberg/pull/34423/commits/5dc8719f9ecf1d31e7515393d50dc50f9fed2cfd)

File: docs/contributors/code/testing-overview.md
Comment (ciampo) [reply docs/contributors/code/testing-overview.md]: Agreed. Done in [`e5de664` (#34423)](https://github.com/WordPress/gutenberg/pull/34423/commits/e5de6647a9dfc4912aac29edbb2f0ca3ae6b6524)

**Comment by ciampo** [comment]
> In addition to `toMatchDiffSnapshot` should there be a blurb about `toMatchStyleDiffSnapshot`?

Absolutely! I've added a paragraph about it in [`3f4ef39` (#34423)](https://github.com/WordPress/gutenberg/pull/34423/commits/3f4ef39ccb8478a1a5e4b9265d4f89da8750fb35).

@gziolo @sarayourfriend @diegohaz all feedback so far should have been tackled, let me know if there's anything else you'd like to see amended!

File: docs/contributors/code/testing-overview.md
Comment (ciampo) [reply docs/contributors/code/testing-overview.md]: > We can also add an ESLint rule that errors when someone adds enzyme in the test file (we would have to add exceptions in existing test files though).

Makes sense, I'll try to fit it in in the next days

File: docs/contributors/code/testing-overview.md
Comment (gziolo) [inline docs/contributors/code/testing-overview.md]: Nit: would it be okay to update to: 
```suggestion
	const { container: spacer } = render( <Spacer /> );
	const { container: spacerWithMargin } = render( <Spacer margin={ 5 } /> );
	expect( spacerWithMargin.firstChild ).toMatchStyleDiffSnapshot(
		spacer.firstChild
	);
```

It might be only me, but I like to read tests like sentences 😅

File: docs/contributors/code/testing-overview.md
Comment (ciampo) [reply docs/contributors/code/testing-overview.md]: Tracked in #34437

**Review by gziolo — APPROVED** [review]

**Comment by gziolo** [comment]
You can wait for @diegohaz for his feedback because he is more experienced with testing components 😄

**Review by diegohaz — APPROVED** [review]
This looks great to me! Thanks @ciampo! 🚀

**Pull request merged**
